rio_tiler/reader.py

```diff
@@ -202,127 +296,121 @@ def part(
+            height, width = _get_width_height(max_size, vrt_height, vrt_width)
 
-    if buffer:
         height = height or vrt_height
         width = width or vrt_width
 
-        # Get output resolution
-        x_res = (bounds[2] - bounds[0]) / out_width
-        y_res = (bounds[3] - bounds[1]) / out_height
-
-        # apply buffer to bounds
-        bounds = (
-            bounds[0] - x_res * buffer,
-            bounds[1] - y_res * buffer,
-            bounds[2] + x_res * buffer,
-            bounds[3] + y_res * buffer,
-        )
-
-        # new output size
-        out_height += int(buffer * 2)
-        out_width += int(buffer * 2)
-
-        # re-calculate the transform given the new bounds, height and width
-        vrt_transform, vrt_width, vrt_height = get_vrt_transform(
-            src_dst, bounds, out_height, out_width, dst_crs=dst_crs
-        )
+        if buffer:
+            bounds, height, width = _apply_buffer(buffer, bounds, height, width)
+            # re-calculate the transform given the new bounds, height and width
+            vrt_transform, vrt_width, vrt_height = get_vrt_transform(
+                src_dst, bounds, height, width, dst_crs=dst_crs
+            )
 
-    window = windows.Window(col_off=0, row_off=0, width=vrt_width, height=vrt_height)
-
-    if padding > 0 and not is_aligned(src_dst, bounds, out_height, out_width, dst_crs):
-        vrt_transform = vrt_transform * Affine.translation(-padding, -padding)
-        orig_vrt_height = vrt_height
-        orig_vrt_width = vrt_width
-        vrt_height = vrt_height + 2 * padding
-        vrt_width = vrt_width + 2 * padding
-        window = windows.Window(
-            col_off=padding,
-            row_off=padding,
-            width=orig_vrt_width,
-            height=orig_vrt_height,
-        )
+        if padding > 0 and not is_aligned(src_dst, bounds, bounds_crs=dst_crs):
+            vrt_transform = vrt_transform * Affine.translation(-padding, -padding)
+            window = windows.Window(
+                col_off=padding, row_off=padding, width=vrt_width, height=vrt_height
+            )
+            vrt_height = vrt_height + 2 * padding
+            vrt_width = vrt_width + 2 * padding
 
-    vrt_options = vrt_options or {}
-    vrt_options.update(
-        {
+        vrt_params = {
             "crs": dst_crs,
             "transform": vrt_transform,
             "width": vrt_width,
             "height": vrt_height,
         }
-    )
+        if vrt_options:
+            vrt_params.update(**vrt_options)
+
+        return read(
+            src_dst,
+            indexes=indexes,
+            width=width,
+            height=height,
+            window=window,
+            nodata=nodata,
+            vrt_options=vrt_params,
+            resampling_method=resampling_method,
+            force_binary_mask=force_binary_mask,
+            unscale=unscale,
+            post_process=post_process,
+        )
+
+    # else no re-projection needed
```

Comment:
👇 is the new Simple dataset Read. Sadly, because of buffer and padding it feels we are duplicating a lot of code ... but I couldn't find a better way!

Reply:
I think we could deduplicate some code but that would require refactoring `read` which doesn't seem like a good idea.  The `if...else` here is easy enough to follow so I don't mind the code duplication too much.